include/lo2s/ringbuf.hpp

```diff
@@ -0,0 +1,271 @@
+/*
+ * This file is part of the lo2s software.
+ * Linux OTF2 sampling
+ *
+ * Copyright (c) 2024,
+ *    Technische Universitaet Dresden, Germany
+ *
+ * lo2s is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "lo2s/error.hpp"
+#include <atomic>
+#include <cassert>
+#include <cstddef>
+#include <cstring>
+#include <string>
+#include <system_error>
+
+extern "C"
+{
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+}
+
+namespace lo2s
+{
+
+// To resolve possible ringbuf format incompatibilities
+#define RINGBUF_VERSION 1
+
+struct ringbuf_header
+{
+    uint64_t version;
+    uint64_t size;
+    std::atomic_uint64_t head;
+    std::atomic_uint64_t tail;
+};
+
+class Mmap
+{
+public:
+    Mmap() : addr_(nullptr), size_(0)
+    {
+    }
+
+    Mmap(Mmap&) = delete;
+    Mmap& operator=(Mmap&) = delete;
+
+    Mmap(Mmap&& other)
+    {
+        std::swap(addr_, other.addr_);
+        std::swap(size_, other.size_);
+    }
+    Mmap& operator=(Mmap&& other)
+    {
+        std::swap(addr_, other.addr_);
+        std::swap(size_, other.size_);
+        return *this;
+    }
+    Mmap(int fd, size_t size, size_t offset, void* location = nullptr) : size_(size)
+    {
+        if (location == nullptr)
+        {
+            addr_ = mmap(nullptr, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, offset);
+        }
+        else
+        {
+            addr_ =
+                mmap(location, size, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_FIXED, fd, offset);
+        }
+        if (addr_ == MAP_FAILED)
+        {
+            throw_errno();
+        }
+    }
+
+    template <typename T>
+    T as()
+    {
+        return reinterpret_cast<T>(addr_);
+    }
+
+    ~Mmap()
+    {
+        munmap(addr_, size_);
+    }
+
+private:
+    void* addr_ = nullptr;
+    size_t size_;
+};
```

Comment:
Extract this as a separate PR and replace `EventReader::init_mmap()` with it?

Reply:
At the very least, extract into separate file

Reply:
https://github.com/tud-zih-energy/lo2s/pull/341